cw_decred/lib/wallet.dart

```diff
@@ -0,0 +1,604 @@
+    if (watchingOnly) {
+      throw "a watching only wallet cannot sign";
+    }
+    var addr = address;
+    if (addr == null) {
+      addr = libdcrwallet.currentReceiveAddress(walletInfo.name);
+    }
+    if (addr == null) {
+      throw "unable to get an address from unsynced wallet";
+    }
+    return libdcrwallet.signMessageAsync(
+        walletInfo.name, message, addr, _password);
+  }
+
+  List<Unspent> unspents() {
+    final res = libdcrwallet.listUnspents(walletInfo.name);
+    final decoded = json.decode(res);
+    var unspents = <Unspent>[];
+    for (final d in decoded) {
+      final spendable = d["spendable"] ?? false;
+      if (!spendable) {
+        continue;
+      }
+      final amountDouble = d["amount"] ?? 0.0;
+      final amount = (amountDouble * 1e8).toInt().abs();
+      final utxo = Unspent(
+          d["address"] ?? "", d["txid"] ?? "", amount, d["vout"] ?? 0, null);
+      utxo.isChange = d["ischange"] ?? false;
+      unspents.add(utxo);
+    }
+    this.updateUnspents(unspents);
+    return unspents;
+  }
+
+  void updateUnspents(List<Unspent> unspentCoins) {
+    if (this.unspentCoinsInfo.isEmpty) {
+      unspentCoins.forEach((coin) => this.addCoinInfo(coin));
+      return;
+    }
+
+    if (unspentCoins.isEmpty) {
+      this.unspentCoinsInfo.clear();
+      return;
+    }
+
+    final walletID = idPrefix + walletInfo.name;
+    if (unspentCoins.isNotEmpty) {
+      unspentCoins.forEach((coin) {
+        final coinInfoList = this.unspentCoinsInfo.values.where((element) =>
+            element.walletId == walletID &&
+            element.hash == coin.hash &&
+            element.vout == coin.vout);
+
+        if (coinInfoList.isEmpty) {
+          this.addCoinInfo(coin);
+        }
+      });
+    }
+
+    final List<dynamic> keys = <dynamic>[];
+    this.unspentCoinsInfo.values.forEach((element) {
+      final existUnspentCoins =
+          unspentCoins.where((coin) => element.hash.contains(coin.hash));
+
+      if (existUnspentCoins.isEmpty) {
+        keys.add(element.key);
+      }
+    });
+
+    if (keys.isNotEmpty) {
+      unspentCoinsInfo.deleteAll(keys);
+    }
+  }
+
+  void addCoinInfo(Unspent coin) {
+    final newInfo = UnspentCoinsInfo(
+      walletId: idPrefix + walletInfo.name,
+      hash: coin.hash,
+      isFrozen: false,
+      isSending: false,
+      noteRaw: "",
+      address: coin.address,
+      value: coin.value,
+      vout: coin.vout,
+      isChange: coin.isChange,
+      keyImage: coin.keyImage,
+    );
+
+    unspentCoinsInfo.add(newInfo);
+  }
+
+  // walletBirthdayBlockHeight checks if the wallet birthday is set and returns
+  // it. Returns -1 if not.
+  int walletBirthdayBlockHeight() {
+    final res = libdcrwallet.birthState(walletInfo.name);
+    final decoded = json.decode(res);
+    if (decoded["setfromheight"] == true || decoded["setfromtime"] == true) {
+      return -1;
+    }
+    return decoded["height"] ?? 0;
```

Comment:
Ah, this is a little confusing, but if those values are set, it means the birthday is not set yet. This would happen during sync, before we find the birthday block.

Reply:
Added a comment.